packages/server/src/core/bank_rule/iterate_courses.rs

```diff
@@ -45,25 +23,53 @@ impl<'a> BankRuleHandler<'a> {
                         course_status.course.id.clone(),
                     );
                 } else {
+                    #[inline(always)]
+                    fn check_if_replacement(
+                        course_list: &Vec<CourseId>,
+                        courses: &HashMap<CourseId, Course>,
+                        course_status: &mut CourseStatus,
+                        course_id_in_list: &mut Option<String>,
+                        replacements: &HashMap<CourseId, OptionalReplacements>,
+                        replacements_msg: impl Fn(&Course) -> String,
+                    ) {
+                        for course_id in course_list {
+                            if let Some(replacements) = &replacements.get(course_id) {
+                                if replacements.contains(&course_status.course.id) {
+                                    let optional_course = courses.get(course_id);
+                                    course_status.set_msg(replacements_msg(
+                                        optional_course.unwrap_or(&Course {
+                                            id: course_id.clone(),
+                                            ..Default::default()
+                                        }),
+                                    ));
+
+                                    *course_id_in_list = Some(course_id.into());
+                                    break;
+                                }
+                            }
+                        }
+                    }
                     // check if course_status is a replacement for a course in course list
                     let mut course_id_in_list = None;
                     // First try to find catalog replacements
-                    check_if_replacement!(
-                        self,
+                    check_if_replacement(
+                        &self.course_list,
+                        self.courses,
                         course_status,
-                        course_id_in_list,
-                        catalog_replacements,
-                        catalog_replacements_msg
+                        &mut course_id_in_list,
+                        self.catalog_replacements,
```

Comment:
Why do we send ```self.course_list, self.course_id_in_list,self.catalog_replacments``` instead of sending self as an argument?
I think it is better because the function will be more concise.

Reply:
Can't have another borrow of self while it is also borrowed as mutable.
Besides, since it's a specific, private, tailor-made macro-like inlined function, conciseness holds little to no value here.